### PR TITLE
Add timezone to CronJob

### LINKED ISSUE #47202: CronJobs should support timezones
I couldn't find any discussion of this after searching for "cron timezone", "cronjob timezone", or "scheduledjob timezone".

The CronJob spec makes reference to https://en.wikipedia.org/wiki/Cron. That page suggests that cron would respect the timezone for a given user. The controller manager runs in a single time zone under a single user so I can't use different time zones for each job. I have jobs that run based on the schedule of external entities that observe daylight savings time. So, if I define that CronJob in UTC I will be forced to update that job from time to time (generally not something one remembers to do after just losing an hour of sleep).

I see two options for how this support might work in kubernetes:

1. Add some new field to the `CronJobSpec`, like `timezone: "Americas/Chicago"`.
2. Use an extended cron syntax that includes timezone, e.g. `30 6 * * 1 Europe/Stockholm`

### PR DESCRIPTION
**Pull request by iterion**
This PR adds a timezone field to the `CronJobSpec` so that cron jobs can run in given timezones. This is necessary for situations where a user wants to run a cron job in locations that observe daylight savings time.

Fixes #47202

```release-note
Add timezone to CronJob
The timezone given in the CronJobSpec will be used when determining when to schedule the job regardless of the default timezone on the box running the controller manager.
```

### CODE REVIEW

**Comment by iterion** [comment]
also fyi: @janetkuo @soltysh

**Comment by kargakis** [comment]
@kubernetes/sig-apps-pr-reviews

File: pkg/controller/cronjob/utils.go
Comment (kargakis) [inline pkg/controller/cronjob/utils.go]: log this error

Comment (kargakis) [inline pkg/controller/cronjob/utils.go]: This should also be an event. What I am wondering is if this is correct. time.Local means the local time this code runs ie. the time in the location the controller manager runs. Not sure we want that.

Comment (kargakis) [inline pkg/controller/cronjob/utils.go]: So LoadLocation can fail only because the timezone specified is invalid? If so, that should be a validation error probably or we shouldn't run the job at all.

Comment (iterion) [inline pkg/controller/cronjob/utils.go]: Yeah, validation sounds like a good idea. I think it's still possible that the list of time zones could change underneath the running controller process, in that case, we should probably not run the job at all and log an event.

As far as `time.Local`, previously we were just using `time.Now()` which returns the local time. So, in this case, the behavior is the same as previous. I would guess that most cluster operators would be using UTC, but, if they're not, they might expect that jobs are scheduled in the time zone of the controller manager.

Comment (iterion) [inline pkg/controller/cronjob/utils.go]: @kargakis I've pushed some changes that attempt address your concerns.

Comment (kargakis) [inline pkg/controller/cronjob/utils.go]: @iterion looks good, thanks

**Comment by janetkuo** [comment]
@k8s-bot ok to test

**Comment by iterion** [comment]
I'm not quite sure how to troubleshoot the federation-e2e-gce test, could someone point me in the right direction?

**Comment by kargakis** [comment]
> I'm not quite sure how to troubleshoot the federation-e2e-gce test, could someone point me in the right direction?

Doesn't seem related to your changes.

**Comment by iterion** [comment]
Ah, it seems I missed the cutoff for 1.7?

**Comment by kargakis** [comment]
Yeah, code freeze was a couple of weeks ago.

**Comment by iterion** [comment]
Bummer, no worries. Is https://github.com/kubernetes/kubernetes/pull/41901 still planned to go into 1.7? I assume I will need to rebase and port this to v1beta1 once that is merged?

**Comment by kargakis** [comment]
> Bummer, no worries. Is #41901 still planned to go into 1.7? I assume I will need to rebase and port this to v1beta1 once that is merged?

Missed the cut, too.

**Comment by soltysh** [comment]
@iterion hey, sorry I was out for a bit of time. I'll reserve some time tomorrow and will review this. Are you still interested in pushing this forward?

**Comment by iterion** [comment]
Yeah, I’d still like to get it in, I can rebase to get it mergeable again
soon.

**Comment by soltysh** [comment]
Cool, let me do the first pass tomorrow and you can address comments and rebase at one shot.

**Comment by iterion** [comment]
Sounds good, thanks!

File: pkg/apis/batch/types.go
Comment (soltysh) [inline pkg/apis/batch/types.go]: A few suggestions:

1. Personally, I'd go with just string, no pointer. In this case both empty or nil means not set. No need to have a distinction between them.
2. `TimeZone` instead of `Timezone` these are two words. 
3. Make sure to tie this to `Schedule` field, iow. move it right after it and add appropriate comment. This is because at some point in time we're planning on supporting [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601#Times), which has its own time zone information and it this can't interfere with it.

File: pkg/apis/batch/validation/validation_test.go
Comment (soltysh) [inline pkg/apis/batch/validation/validation_test.go]: Please add a valid case as well.

File: pkg/controller/cronjob/cronjob_controller.go
Comment (soltysh) [inline pkg/controller/cronjob/cronjob_controller.go]: 👍

File: pkg/controller/cronjob/cronjob_controller.go
Comment (soltysh) [inline pkg/controller/cronjob/cronjob_controller.go]: Use more descriptive variable names, like tzTime. It makes reading code so much easier and more obvious :)

File: pkg/controller/cronjob/utils.go
Comment (soltysh) [inline pkg/controller/cronjob/utils.go]: Return nil with error, no time. This may cause some errors slip through. When you return nil, you'll get a nasty nil pointer error which will allow tracking what actually went wrong.

**Review by soltysh — CHANGES_REQUESTED** [review]
Overall this looks good to me. I left you some comments.

**Comment by soltysh** [comment]
@erictune || @janetkuo mind having a second look. I'd like to merge this asap, so I can pick it up with the beta promotion for 1.8.

**Comment by iterion** [comment]
Thanks for the review, I should be able to address your comments and rebase sometime tomorrow.

File: pkg/apis/batch/v2alpha1/types.go
Comment (iterion) [inline pkg/apis/batch/v2alpha1/types.go]: @soltysh should I move  TimeZone from here to just below schedule as well? If so, should it continue to be `bytes,8,opt,name=timezone`. It's my understanding that changing the ordering of these in protobuf would break backwards compatibility?

Comment (soltysh) [inline pkg/apis/batch/v2alpha1/types.go]: The protobuf generator should handle it accordingly. The number (here 8) should stay, otherwise this will break backwards compatibility.

**Comment by soltysh** [comment]
Also @kubernetes/api-approvers @kubernetes/sig-api-machinery-api-reviews for api approval

File: pkg/controller/cronjob/utils.go
Comment (iterion) [inline pkg/controller/cronjob/utils.go]: @soltysh was your intention here that I change the return type to `(*time.Time, error)`? Otherwise, as it stands, it would be impossible to `return nil, err`.

**Comment by iterion** [comment]
@soltysh pushed up requested changes, did have one question on a return type.

**Comment by iterion** [comment]
@soltysh sorry to bother, any chance you could give this a review again?

File: pkg/controller/cronjob/cronjob_controller.go
Comment (kargakis) [inline pkg/controller/cronjob/cronjob_controller.go]: Can you include the error in the event message?

**Comment by kargakis** [comment]
/lgtm

Needs a sign-off/approval from @erictune and @soltysh

**Comment by kargakis** [comment]
@iterion is there any issue in the features repo for this?

**Comment by kargakis** [comment]
> @iterion is there any issue in the features repo for this?

@soltysh is this covered by https://github.com/kubernetes/features/issues/19?

File: pkg/controller/cronjob/utils.go
Comment (soltysh) [inline pkg/controller/cronjob/utils.go]: My bad, haven't noticed that. Yeah I guess so.

File: pkg/controller/cronjob/cronjob_controller.go
Comment (soltysh) [inline pkg/controller/cronjob/cronjob_controller.go]: This can be fixed in a followup, as well. Let's try to get this one in.

**Review by soltysh — APPROVED** [review]

File: staging/src/k8s.io/api/batch/v2alpha1/types.go
Comment (soltysh) [inline staging/src/k8s.io/api/batch/v2alpha1/types.go]: Now that https://github.com/kubernetes/kubernetes/pull/41901 is merged, mind adding it also to `batch/v1beta1`. It's part of the beta reqs.

**Comment by soltysh** [comment]
> @soltysh is this covered by kubernetes/features#19?

Yes, it's part of the beta features. 

@iterion it all looks good, we need final api approval from @erictune and this also needs to be added to `batch/v1beta1`, if you manage to push the changes before Eric. If not a followup is fine. 

/retest

**Comment by soltysh** [comment]
@iterion make sure to update all the deps. It looks like not all have been.

**Comment by iterion** [comment]
@soltysh Added it to `batch/v1beta1`. I assume by "update all the deps" you mean to run `hack/update-all.sh`? If so, doing that now and pushing shortly.

**Comment by iterion** [comment]
The bazel fails seem to be related to missing timezone info, I'm not sure exactly how to resolve those issues, any guidance would be appreciated:

```
Unexpected error  open /usr/local/go/lib/time/zoneinfo.zip: no such file or directory
```

**Comment by janetkuo** [comment]
Haven't looked at the failure closely but this might be related: https://github.com/golang/go/issues/20969

**Comment by erictune** [comment]
I'm a little nervous about the `time.LoadLocation` call.

- If there is no time zone database, will the cron controller still work?  I think it should.
- which time zone database gets used?  The one in the container image that the kube-controller-manager runs in?  
- whose job is it to make sure that timezone database is up to date? 

@tallclair any issue with depending on ZONEINFO on the kube-controller-manager's base image?

**Comment by iterion** [comment]
@erictune to try to answer some of your concerns. 

As it's currently written, the cron controller should still work without a time zone database available. Users would simply not be able to create cron jobs with a nonempty `timeZone`. In both places where LoadLocation is called (validation and before scheduling the job) we default to using UTC which should be safe even with zone info missing. In the cases in the unit tests where `timeZone == ""` it seems all have passed on bazel.

The time zone database that is used would depend on a number of scenarios based on the [time.LoadLocation](https://golang.org/pkg/time/#LoadLocation) implementation. But, ultimately it would be sourced from the container/server on which the kube-controller-mananger is running.

Your last question is a bit harder to answer. Ultimately, imo, if a cluster admin needed the most up to date IANA time zone database I think the onus would be on them to keep that up to date as with any dependency on their host OS.

**Comment by soltysh** [comment]
> The time zone database that is used would depend on a number of scenarios based on the time.LoadLocation implementation. But, ultimately it would be sourced from the container/server on which the kube-controller-mananger is running.

That's my understanding and assumption as well. 
Ditto for updating this file being the role of cluster admin.

> As it's currently written, the cron controller should still work without a time zone database available. Users would simply not be able to create cron jobs with a nonempty timeZone. In both places where LoadLocation is called (validation and before scheduling the job) we default to using UTC which should be safe even with zone info missing. In the cases in the unit tests where timeZone == "" it seems all have passed on bazel.

The only question left is how to update k8s infra to have that database installed. Because it'll be giving false positives on tests. I think Tim will be able to help here with this.

**Comment by soltysh** [comment]
Other than that, this LGTM. Will wait for final sign off from Eric, though.

**Comment by soltysh** [comment]
@mikedanese since Tim's out mind giving some info in this manner, wrt to zoneinfo:

> @tallclair any issue with depending on ZONEINFO on the kube-controller-manager's base image?

**Comment by soltysh** [comment]
I've created https://github.com/kubernetes/test-infra/pull/4232 to address this issue. Once that merges we can continue with this one. 

@erictune any further objections other than that for TZ? Assuming we'll get that TZ test related problem fixed, I'd like to get this one in for 1.8.

**Comment by erictune** [comment]
How is error surfaced to user when `timeZone` is non-empty and not supported?  Status or event on the CronJob please.  Crashing == not acceptable in any circumstance.

**Comment by erictune** [comment]
I think users are going to expect this to be field to be portable.  Most fields in Kubernetes APIs are either portable or obviously cluster dependent (like e.g. Ceph volume type).  Nobody is going to expect `timeZone` to be not-portable.   

Therefore, I'd rather that Kubernetes project stood behind it being portable.

So if @tallclair and @kubernetes/sig-release-api-reviews agree that *the base image for gcr.io/kube-controller-manager shall henceforth always contain a recent IANA TZ db*, then I will approve this API change.

**Comment by soltysh** [comment]
> How is error surfaced to user when timeZone is non-empty and not supported? Status or event on the CronJob please. Crashing == not acceptable in any circumstance.

You'll get validation error when submitting one.

> Therefore, I'd rather that Kubernetes project stood behind it being portable.

I'm not entirely sure what you mean with that? Can you elaborate? 
Personally, I've always assumed the cronjob trigger times should be TZ-dependant on the master, if that's what you're leaning towards this. If so, we need to update the cronjob features not to have TZ ever. But I'll defer to requested reviewers before making final call.

**Comment by ixdy** [comment]
FYI, in case it matters: the kube-controller-manager image is currently [based on busybox](https://github.com/kubernetes/kubernetes/blob/6f4a15f92295b6b99c76ef00c2d4e28f0fd0898d/build/common.sh#L96), so you'd need to download/save a zone info file from somewhere to add to this image (or mount it into the docker container at runtime, which seems probably worse).

**Comment by bgrant0607** [comment]
If this is introducing a new dependency on the controller-manager's image, please ensure that is documented somewhere and mentioned in the release note.

**Comment by iterion** [comment]
I'm going to wait until https://github.com/kubernetes/test-infra/pull/4232 merges before rebasing this PR.

**Comment by mattfarina** [comment]
Would be nice to get this done in the 1.9 timeframe.

File: pkg/controller/cronjob/cronjob_controller.go
Comment (kargakis) [reply pkg/controller/cronjob/cronjob_controller.go]: Please, address this comment alongside the rebase. Things like this tend to fall through to obscurity if there is no accompanying issue and this change is small enough that I don't think it warrants an issue.

**Comment by mattfarina** [comment]
I agree with @erictune that this needs to be portable. That's what people will expect. I don't think it should be too much to require the TZ data as a dependency.

**Comment by rphillips** [comment]
The TZ data is updated frequently from IANA, sometimes monthly. Since the TZ data would be baked into a released image, it could potentially be out of date at some point in the future without the user knowing. Is this concern something that would be documented?

**Comment by iterion** [comment]
I'd love to get rolling again on this one. I think we're still just waiting on https://github.com/kubernetes/test-infra/pull/4232.

**Comment by BenTheElder** [comment]
I think time zone aware cron sounds useful and would potentially use it myself but I have some concerns:

- I don't see any responses regarding the controller image, which seems like a problem.

- How will users know how up to date the TZData they are writing their cron against is?

- When is the controller image data updated? 

- Where are the tzdata considerations documented?
  - Will I know how up to date a release is and have a warning that other clusters may have varyingly stale data?

- How easily can I update the image data myself and is this documented?

I think perhaps some more discussion of the tzdata management is warranted, at the very least the controller image probably needs work.

**Comment by iterion** [comment]
@BenTheElder thanks for noting your concerns, I agree that these things might need better documentation. It seems like the best option for everyone might be ensuring that tzdata is always kept fresh. But, given the wide number of methods for deployments, I'm not sure if that's something we can reasonably guarantee? Which means the onus would be on the cluster operator to ensure they stay up to date.

> How will users know how up to date the TZData they are writing their cron against is?
Could we use events to surface this? Or, perhaps log something when controller-manager boots?

> When is the controller image data updated?
Unless we want to maintain some separate process to ensure this stays up to date and push new releases when tzdata is updated, then probably the best we can guarantee is every release. Alternatively, users could mount the host OS's tzdata into the controller image.

> Where are the tzdata considerations documented?
They're not, but I'm happy to do so, I'm just not sure where it should go.

> How easily can I update the image data myself and is this documented?
Also not documented, also unsure about where it should go?

**Comment by soltysh** [comment]
> How will users know how up to date the TZData they are writing their cron against is?
> 
> Could we use events to surface this? Or, perhaps log something when controller-manager boots?

Event seems reasonable, we're using it also for informing about problems with TZ so that sgtm.

> Where are the tzdata considerations documented?
> 
> They're not, but I'm happy to do so, I'm just not sure where it should go.

That I think should be documented in the cronjob part, here: https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/

As for the information about how to update image, I'm guessing somewhere in cluster administration section. Not sure where, but if you submit a PR docs team will help you figure out where it should go. It can easily be moved afterwards, as well. 

> I'd love to get rolling again on this one. I think we're still just waiting on kubernetes/test-infra#4232.

Per @ixdy suggestion I'm proposing to inject `getCurrentTimeInZone` in tests so that it does not rely on anything on the image. This would ensure the test is stable and much more predictable. @iterion can you address this in here? 

> Would be nice to get this done in the 1.9 timeframe.

I apologise, I got slightly overwhelmed with the amount of work over the past few months. Let's get this going for 1.10? /cc @mattfarina

**Comment by tamalsaha** [comment]
@soltysh, we would really appreciate if this can be done in 1.9 cycle. :pray:

**Comment by soltysh** [comment]
@tamalsaha sorry, the gates for 1.9 already closed. Only high prio bugs are getting in.

**Comment by iterion** [comment]
@soltysh thanks for the review. I've rebased and fixed up some changes to get this mergeable. Could you ok this to test so we can see where we stand? I ran unit tests locally and they seem fine, but I'd like to see how e2e tests are doing so I can try to stub things out there.

**Comment by BenTheElder** [comment]
FYI this [looks like it needs another rebase](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/47266/pull-kubernetes-verify/68818/) and probably hack/update-all (sampled the failures).

**Comment by sdminonne** [comment]
/sub
@soltysh I'm late on this topic (sorry)

**Comment by smarterclayton** [comment]
The biggest unaddressed issue here is portability - making the Kubernetes API dependent on particular interpretations of time zone data.  This seems like a "hard problem" in the sense that time zones are human defined, change frequently, and are not reliably updated.  It means old cron job definitions may stop working at some point in the future.  It means a cron job that runs on one cluster (GKE) would behave differently from one running in a bare metal environment.  Are these desirable properties of the Kube API?

I'm slightly against taking any dependency on something as ephemeral and prone to change as time zones.  It puts more of a burden on clients and users, but I'd be more concerned that a cron job behaves differently across clusters or on different nodes or different Kube releases, thus potentially leading to long term issues.

I'm leaning against anything except UTC times in the cron job format, even though I understand why from a usability issue this is a problem for users.  I'd need convincing that we can make time zones consistent across clusters and versions of Kubernetes.

**Comment by erictune** [comment]
This is a "time offset" rather than a "time zone", I guess.

**Comment by soltysh** [comment]
I'm personally a bit reluctant to add any time-related changes. None of the resources we create has that notion, so a user may expect the times coming from the server are always the same (all UTC). With time zones or time offsets we're breaking that contract and this might confuse users.

**Comment by smarterclayton** [comment]
Fixed offset does seem at least more understandable (since I'm less worried about us moving off a 24 hour day).  Fixed offset wouldn't handle DST switches any better than UTC, but would be slightly less confusing than pure UTC.  Same question about "does it improve the usability enough"

**Comment by iterion** [comment]
I originally embarked on this endeavor to ease the pain of some DST issues, so losing that would make this feature rather pointless for my original use case. I don't think an offset would really improve the usability enough, as @smarterclayton suggests.

It seems users could simply write a small controller to update their cronjobs when they pass through DST. A simple annotation (`update-for-dst: true`) and a controller would just as easily solve my original use case and not place any additional burden on Kubernetes or cluster admins to maintain tz data.

**Comment by whereisaaron** [comment]
I can certainly see why Kubernetes would want to dodge the complexities of TZ handling, but then managing CronJob expressions in UTC in a TZ+DST world is painful for humans 😄. Many (most?) CronJobs that aren't 'every X minutes' are targeted at the human TZ UX. anyway. Who are we targeting here, computer convenience or human convenience? :smile:

I can see how a TZ aware controller would work @iterion, but it many clusters it would be updating every single CronJob in the cluster in the same minute, twice a year. A nice spike for `etcd` 😄 

I think you would still need to annotate the CronJobs with the current timezone too? Otherwise the controller instance(s) won't know which ones have already been processed (and might double-adjust). Also - If a new CronJob is created, which timezone and DST state is the offset currently in? And so does it need adjusting?

**Comment by soltysh** [comment]
Based on [this comment](https://github.com/kubernetes/kubernetes/issues/47202#issuecomment-360820586) I think we can close this pr.

**Pull request closed without merging**
